frontend/ui/widget/imageviewer.lua

```diff
@@ -0,0 +1,313 @@
+local Button = require("ui/widget/button")
+local UIManager = require("ui/uimanager")
+local Screen = require("device").screen
+local Device = require("device")
+local Geom = require("ui/geometry")
+local Event = require("ui/event")
+local Font = require("ui/font")
+local logger = require("logger")
+local _ = require("gettext")
+local T = require("ffi/util").template
+local Blitbuffer = require("ffi/blitbuffer")
+
+--[[
+Display image with some simple manipulation options
+]]
+local ImageViewer = InputContainer:new{
+    -- Allow for providing same different input types as ImageWidget :
+    -- a path to a file
+    file = nil,
+    -- or an already made BlitBuffer (ie: made by Mupdf.renderImageFile())
+    image = nil,
+    -- whether provided BlitBuffer should be free(), normally true
+    -- unless our caller wants to reuse it's provided image
+    image_disposable = true,
+
+    fullscreen = false, -- false will add some padding around widget (so footer can be visible)
+    with_title_bar = true,
+    title_text = _("Viewing image"), -- default title text
+
+    width = nil,
+    height = nil,
+    stretched = true, -- start with image stretched (Best fit)
+    rotated = false,
+    -- we use this global setting for rotation angle to have the same angle as reader
+    rotation_angle = DLANDSCAPE_CLOCKWISE_ROTATION and 90 or 270,
+
+    title_face = Font:getFace("tfont", 22),
+    title_padding = Screen:scaleBySize(5),
+    title_margin = Screen:scaleBySize(2),
+    image_padding = Screen:scaleBySize(2),
+    button_padding = Screen:scaleBySize(14),
+
+    -- Reference to current ImageWidget instance, for cleaning
+    _image_wg = nil,
+}
+
+function ImageViewer:init()
+    if Device:hasKeys() then
+        self.key_events = {
+            Close = { {"Back"}, doc = "close viewer" }
+        }
+    end
+    if Device:isTouchDevice() then
+        self.ges_events = {
+            Tap = {
+                GestureRange:new{
+                    ges = "tap",
+                    range = Geom:new{
+                        x = 0, y = 0,
+                        w = Screen:getWidth(),
+                        h = Screen:getHeight(),
+                    }
+                },
+            },
+            Swipe = {
+                GestureRange:new{
+                    ges = "swipe",
+                    range = Geom:new{
+                        x = 0, y = 0,
+                        w = Screen:getWidth(),
+                        h = Screen:getHeight(),
+                    }
+                },
+            },
+        }
+    end
+    self:update()
+end
+
+function ImageViewer:_clean_image_wg()
+    -- To be called before re-using / not needing self._image_wg
+    -- otherwise ressources used by its blitbuffer won't be freed
+    if self._image_wg then
+        logger.dbg("ImageViewer:_clean_image_wg()")
+        self._image_wg:free()
+        self._image_wg = nil
+    end
+end
+
+function ImageViewer:update()
```

Comment:
Not sure how.
Here, we're mostly just instantiating UI basic elements to build the whole UI, and they need to be local to be packed into other elements. It's indeed long when the widget is not a simple one (same in `DictQuickLookup:update()` ). I think splitting would make following the building logic less easy.

It's also long because I let in some titlebar stuff that I used first, and that i made optional cause it was best to have more room for the image. I thought I'd let that in for others who may one day want to implement viewing images from directories, where the title could be used to show the filename of the currently displayed image.